src/notebooks/controllers/controllerPreferredService.ts

```diff
@@ -0,0 +1,318 @@
+import { noop } from '../../platform/common/utils/misc';
+import { IInterpreterService } from '../../platform/interpreter/contracts';
+import { traceError, traceInfo, traceInfoIfCI, traceVerbose } from '../../platform/logging';
+import { PythonEnvironment } from '../../platform/pythonEnvironments/info';
+import { sendTelemetryEvent } from '../../telemetry';
+import { getTelemetrySafeLanguage } from '../../telemetry/helpers';
+import {
+    IControllerDefaultService,
+    IControllerLoader,
+    IControllerPreferredService,
+    IControllerRegistration,
+    IVSCodeNotebookController,
+    PreferredKernelExactMatchReason
+} from './types';
+
+/**
+ * This class keeps track of selected controllers
+ */
+@injectable()
+export class ControllerPreferredService implements IControllerPreferredService, IExtensionSingleActivationService {
+    private preferredControllers = new Map<NotebookDocument, IVSCodeNotebookController>();
+    private get isLocalLaunch(): boolean {
+        return this.serverConnectionType.isLocalLaunch;
+    }
+    constructor(
+        @inject(IControllerRegistration) private readonly registration: IControllerRegistration,
+        @inject(IControllerLoader) private readonly loader: IControllerLoader,
+        @inject(IControllerDefaultService) private readonly defaultService: IControllerDefaultService,
+        @inject(IInterpreterService) private readonly interpreters: IInterpreterService,
+        @inject(IVSCodeNotebook) private readonly notebook: IVSCodeNotebook,
+        @inject(IDisposableRegistry) readonly disposables: IDisposableRegistry,
+        @inject(IKernelFinder) private readonly kernelFinder: IKernelFinder,
+        @inject(IPythonExtensionChecker) private readonly extensionChecker: IPythonExtensionChecker,
+        @inject(ServerConnectionType) private readonly serverConnectionType: ServerConnectionType
+    ) {}
+    public async activate() {
+        // Sign up for document either opening or closing
+        this.notebook.onDidOpenNotebookDocument(this.onDidOpenNotebookDocument, this, this.disposables);
+        // If the extension activates after installing Jupyter extension, then ensure we load controllers right now.
+        this.notebook.notebookDocuments.forEach((notebook) => this.onDidOpenNotebookDocument(notebook));
+    }
+    public async computePreferred(
+        document: NotebookDocument,
+        serverId?: string | undefined
+    ): Promise<{
+        preferredConnection?: KernelConnectionMetadata | undefined;
+        controller?: IVSCodeNotebookController | undefined;
+    }> {
+        traceInfoIfCI(`Clear controller mapping for ${getDisplayPath(document.uri)}`);
+        const loadControllersPromise = this.loader.loaded;
+        // Keep track of a token per document so that we can cancel the search if the doc is closed
+        const preferredSearchToken = new CancellationTokenSource();
+        const disposable = this.notebook.onDidCloseNotebookDocument(
```

Comment:
We'll end up creating multiples of these for each `computePreferred` method.
I think we could instead just create a common handler in the `active` method and store the cancellation tokesn in a dictionary.
& then the handler is invoked we look for items in that dictionary (keyed by document) & then cancel that token.

Reply:
That does seem better. But I didn't write any of this. This is all existing code. I guess that seems harmless enough, but I was trying to not change any of the logic (or as little of the logic as possible).